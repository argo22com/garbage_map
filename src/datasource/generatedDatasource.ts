import { Container, ContainerType, Spot } from "datasource/index";
import rawSpots from "./generated/spots.json";
import rawContainers from "./generated/trashes.json";

export async function getMappedData(): Promise<Spot[]> {
  // map raw containers json to correct type
  const containerMap = new Map<string, Container>(
    rawContainers.map((rawContainer) => {
      return [
        rawContainer.uid,
        {
          uid: rawContainer.uid,
          clearDay: rawContainer.clear_day,
          type: rawContainer.type as ContainerType,
        },
      ];
    })
  );

  // map raw spots json to correct type
  return rawSpots.map((rawSpot) => {
    return {
      uid: rawSpot.uid,
      address: rawSpot.location.address_simple,
      location: {
        latitude: rawSpot.location.location_latitude,
        longitude: rawSpot.location.location_longitude,
      },
      containers: rawSpot.trashes
        .filter((uid) => containerMap.has(uid))
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        .map((uid) => containerMap.get(uid)!),
    };
  });
}
