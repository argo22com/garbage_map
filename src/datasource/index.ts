import { getMappedData } from "./generatedDatasource";
import { Filters, Spot } from "./types";

function applyFilters(filters: Filters, spots: Spot[]) {
  const containerTypes = filters.containerTypes;
  if (containerTypes) {
    spots = spots.filter((spot) =>
      containerTypes.every((type) =>
        spot.containers.find((container) => container.type === type)
      )
    );
  }
  return spots;
}

export async function getSpots(filters: Filters = {}): Promise<Spot[]> {
  // TODO: implement datasource selection once we need it
  let spots = await getMappedData();
  spots = applyFilters(filters, spots);

  return spots;
}
