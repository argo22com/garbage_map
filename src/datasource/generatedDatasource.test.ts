import { getMappedData } from "datasource/generatedDatasource";

test("properly maps spots and containers", () => {
  // mock spots - we don't care about actual data, but about the structure, which has to remain the same
  jest.mock("./generated/spots.json", () => [
    {
      id: 1,
      location: {
        address_simple: "A. Trägera 266/113",
        location_latitude: 49.005944075539205,
        location_longitude: 14.476660503581881,
        location_origin_x: 1162728.497,
        location_origin_y: 755191.194,
      },
      trashes: ["t1", "t2"],
      uid: "0cbcb8b2-7de5-4603-bc9a-da5b699d9172",
    },
  ]);
  // mock trashes - we don't care about actual data, but about the structure, which has to remain the same
  jest.mock("./generated/trashes.json", () => [
    {
      capacity: "1 x 2500 l, spodní výsyp",
      clear_day: "úterý",
      clear_frequency: "1 x týdně",
      mark: "0cbcb8b2-7de5-4603-bc9a-da5b699d9172",
      type: "Papír",
      uid: "t1",
    },

    {
      capacity: "1 x 2500 l, spodní výsyp",
      clear_day: "pátek, úterý",
      clear_frequency: "2 x týdně",
      mark: "0cbcb8b2-7de5-4603-bc9a-da5b699d9172",
      type: "Kovy",
      uid: "t2",
    },
  ]);
  getMappedData().then((data) => expect(data).toMatchSnapshot());
});
