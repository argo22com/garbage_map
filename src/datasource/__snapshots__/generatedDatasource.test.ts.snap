// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`properly maps spots and containers 1`] = `
Array [
  Object {
    "address": "Plzeňská 608/71, naproti",
    "containers": Array [
      Object {
        "clearDay": "středa",
        "type": "Plast",
        "uid": "1",
      },
    ],
    "location": Object {
      "latitude": 48.990644999,
      "longitude": 14.463385,
    },
    "uid": "Plzeňská 608/71, naproti",
  },
]
`;
